src/librustc/middle/const_eval.rs

```diff
@@ -1104,10 +854,20 @@ pub fn eval_const_expr_partial<'tcx>(tcx: &ty::ctxt<'tcx>,
 
           let mut call_args = NodeMap();
           for (arg, arg_expr) in decl.inputs.iter().zip(args.iter()) {
+              let arg_hint = match ty_hint {
+                  ExprTypeChecked => ExprTypeChecked,
+                  UncheckedExprNoHint | UncheckedExprHint(_) => {
+                      if let Some(hint) = tcx.ast_ty_to_ty_cache.borrow().get(&arg.ty.id) {
+                          UncheckedExprHint(hint)
```

Comment:
This does not seem correct -- you are just taking the type from the target function, but the target function might be generic, in which case this type hint would be referencing unsubstituted type parameters. The correct thing to do would be to get the type of `callee` and extract the fn types from there. I'm not 100% sure how to write a test that exposes this problem, but it seems like it should be possible: invoke some generic `const fn`, specifying the values for its type parameters, in some context where the type hint is needed.